### ISSUE TITLE
Testing HTTPS

### ISSUE THREAD

**Issue by tmanternach** [scripting/drummerSupport#78]
I saw on Dave's Drummer change notes that he posted about HTTPS today. Here is my experience with making this work.

	
Preface: I am going to assume you already have (or know how to create) an SSL certificate for your domain. If you need a nudge, Google your web server (Apache/nginx) and "Lets Encrypt" for tons of instructions on how to deploy a free one. All of my experience is with a DigitalOcean VPN running Ubuntu and nginx. You can visit my working site [here](https://trevormanternach.com/). 

**I'd love to help anyone else work through this!**

1. You need to host the 3 template files listed in the Drummer docs yourself. These are index.html, styles.css, and code.js. They are discussed in the Drummer docs [here](http://docserver.scripting.com/drummer/blogging.opml#1628776805000).
2. These files need to be accessible on the internet via https. I did this using a sub-domain (so you technically would need 2 SSL certificates). I couldn't figure out a clean way to server these from the same domain I am using for Drummer.
3. **You need to edit your index.html template file and replace all the references to fargo.io and scripting.com with the s3.amazonaws.com/fargo.io and s3.amazonaws.com/scripting.com equivalents.** [Here is what my index.html file looks like after those edits](https://gist.github.com/tmanternach/d0ebd6cd0b33e5faa23658baad37df33). The reason for this is so these files get downloaded via https in your browser. If not, modern browsers don't allow for "mixed content" (the main page loaded via https but script/resources loaded via http).
4. Once your index.html is accessible via https, you need to add a "urlTemplate" OPML header to your blog.opml. This is done by clicking File->Edit OPML Headers in Drummer and clicking the plus button. The value of mine, for example, is "https://static.trevormanternach.com/index.html"
5. You need to configure your web server properly to 1) proxy your main blog to your oldschool.script.com url and 2) serve your 3 template files via https. [Here is my nginx config](https://gist.github.com/tmanternach/5f0fe19b2da0175cc086e77083199987) for doing all of this.

Cheers; have a good weekend everyone!

**Comment by scripting** [comment]
I added a link to this item from the [change note](http://drummer.scripting.com/?url=http://drummer.scripting.com/davewiner/drummer/changeNotes.opml&cursor=1634330202000) about HTTP.

**Comment by sabre23t** [comment]
> You can visit my working site [here](https://trevormanternach.com/).

I just reviewed your blog, tmanternach. It doesn't have the twitter sign on icons, and no retweet buttons for the post. Was that intentional or limitation in https mode?

**Comment by tmanternach** [comment]
@sabre23t, good catch! Not intentional, no. I'll take a look at this tomorrow and report back.

**Comment by am1t** [comment]
@tmanternach This is brilliant. I was able to set my blog behind HTTPS too. I also managed to make the same domain serve the static files. Here's the relevant [nginx.conf](https://gist.github.com/am1t/d3ac570fbf1143a78ccbb41ab4dbbc47) file. Thank you for attempting this and documenting this!

**Comment by tmanternach** [comment]
@am1t Looks great. Your Twitter sign on icon is even working. I need to see what I messed up on my site for that and the reblog buttons.

Good work with the aliases. I had tried that the other day, but didn't try with a subfolder like you did and I ran into issues.

**Comment by tmanternach** [comment]
@jackbaty and @am1t, I think I figured out my issues! I had somehow managed to use some _really_ different template files. I'm not even going to pretend to understand how I did that, but I grabbed new ones from that drummerCms repo and everything looks happier now.

I've updated my index.html gist from above to reflect my now corrected template.

Lastly, @jackbaty, to comment on your update about links not being https, I think that isn't a huge deal. Modern browsers try pretty heavily to prefer https, and your web server looks to be properly redirecting a true http:// request to https://

```
trevormanternach@Trevors-MacBook-Pro ~> curl "http://rudimentarylathe.org/rss.xml" -I
HTTP/1.1 301 Moved Permanently
Server: nginx/1.18.0 (Ubuntu)
Date: Sat, 16 Oct 2021 13:16:45 GMT
Content-Type: text/html
Content-Length: 178
Connection: keep-alive
Location: https://rudimentarylathe.org/rss.xml
```

**Comment by scripting** [comment]
BTW, the template files are already accessible over HTTPS.

If you use these URLs, I will be able to fix any bugs that come up in them and have those changes go through to the HTTPS sites.

1. https://s3.amazonaws.com/scripting.com/code/drummercms/templates/minimal/styles.css

2. https://s3.amazonaws.com/scripting.com/code/drummercms/templates/minimal/code.js

3. https://s3.amazonaws.com/scripting.com/code/drummercms/templates/minimal/index.html

**Comment by jackbaty** [comment]
At the moment, the About tabs are not loading content on any of the blogs from this thread that are otherwise working well under https, mine included (https://rudimentarylathe.org/?tab=about). See also Amit's at https://ol.amitgawande.com/?tab=about. 

I'm posting this in the https thread, because it feels https related due to this error that appears on the console (in latest Safari, macOS):

```
[Error] XMLHttpRequest cannot load http://counters.scripting.com/counter?group=drummer&referer=https%3A%2F%2Frudimentarylathe.org%2F%3Ftab%3Dblog&url=https%3A%2F%2Frudimentarylathe.org%2F%3Ftab%3Dabout due to access control checks.

	send (jquery-1.9.1.min.js:5:17824)
	ajax (jquery-1.9.1.min.js:5:13495)
	readHttpFile (code.js:411)
	hitCounter (code.js:612)
	(anonymous function) (code.js:876)
	startTabsIfHomePage (code.js:145)
	startup (code.js:871)
	(anonymous function) (rudimentarylathe.org:510)
	c (jquery-1.9.1.min.js:3:7979)
	fireWith (jquery-1.9.1.min.js:3:8776)
	ready (jquery-1.9.1.min.js:3:3394)
	H (jquery-1.9.1.min.js:3:817)
```

I've done the usual hard reload, multiple browser testing with the same result.

**Comment by papascott** [comment]
@dcoletta I was following @tmanternach's [advice at the top of this issue](https://github.com/scripting/drummerSupport/issues/78#issue-1027882636) to change the template to link to Dave's files directly on S3. Those are Dave's actual S3 buckets, so we'll catch any changes he makes. 

There are a couple of http requests inside the client-side scripts (that's why the About pages aren't working), but that's a task for another day. 😄

**Comment by jackbaty** [comment]
@papascott You're saying that the failing request is generated from somewhere in the template script (code.js)? I didn't see it there, but if this is something we (as users) can modify ourselves I'm all ears :).

**Comment by dcoletta** [comment]
With that extra clue I've got https://cannabiscto.com working. Thanks!

**Comment by tmanternach** [comment]
> At the moment, the About tabs are not loading content on any of the blogs from this thread that are otherwise working well under https, mine included (https://rudimentarylathe.org/?tab=about). See also Amit's at https://ol.amitgawande.com/?tab=about.

@jackbaty, you are on to something here. I think the console message you posted is unrelated, though (the one you posted is probably a difference problem, btw). When I browse to my About tab via https, I get a slightly different message about failing to load a request to `http://httpproxy.scripting.com/httpReadUrl`.

It looks like Dave has open-sourced the software running at that url in [this repo](https://github.com/scripting/httpProxy). Just as a proof of concept, I am going to try to run this node program behind an HTTPS endpoint and edit my code.js file (starting at line 426 for those following along) to replace the `httpproxy.scripting.com` url with my own. This wouldn't be any sort of production fix, just me mostly trying to see what is possible here.

**Comment by dcoletta** [comment]
I added a favicon to my https-via-Caddy setup. It was non-trivial, because I decided to put the icon resources into PagePark, so they had to be served from the tmpl subdomain. But since the site manifest was also in the tmpl subdomain, I had to enable CORS in order to `<link>` to it.

This worked, but I'm probably missing something. If there is support for favicons in Drummer, then there's probably a way to do this that doesn't involve hosting my own site manifest. 

Or maybe there is an option in Caddy's `reverse_proxy` command that allows me to write a regular expression for URLs not to proxy.

**Comment by jackbaty** [comment]
@tmanternach You're absolutely right, I pasted the wrong error from the console, making my report nearly useless, sorry. I've edited the comment with the correct error. Thanks for pointing that out.

**Comment by tmanternach** [comment]
@jackbaty, I got this to work ([proof here](https://trevormanternach.com/?tab=about)), but it is a huge mess that I don't think is sustainable.

I am not actually running the httpproxy node app myself. I am just proxying it (much like the blog itself) through an HTTPS URL (`https://static.trevormanternach.com/httpReadUrl` proxies to `http://httpproxy.scripting.com/httpReadUrl`).

I then had to download (and host myself) [a separate code.js](http://s3.amazonaws.com/scripting.com/code/includes/basic/code.js) file. My version of that file is [here](https://static.trevormanternach.com/basic-code.js). I replaced the script src in index.html template (line 19 on mine) to replace "my" version with the original. I then edited "my" basic-code.js file and replaced line 426 with my proxy URL for httpReadUrl.

TLDR: Giant kludge, would not recommend.

**Comment by scripting** [comment]
what's the question, as concisely as possilble?

**Comment by dcoletta** [comment]
As a Drummer blogger, I'd like my https-hosted Drummer blog to be able to load the content of a tab without a mixed-content error.

**Comment by scripting** [comment]
ok cool! that's the best thing possible. I already have far too many things to do! :-)

**Comment by am1t** [comment]
@scripting That looks to be the case. Here's the console log again while loading the About page (mainly for sites deployed on HTTPS) - `XMLHttpRequest cannot load http://httpproxy.scripting.com/httpReadUrl?url=http%3A%2F%2Fdrummer.scripting.com%2F_am1t%2Fabout.opml&type=text%2Fplain due to access control checks`. 

I believe if it is to be solved, it will need to be solved in one of the `util` packages -- mainly this function `readHttpFileThruProxy` from whatever I would figure out.

**Comment by tmanternach** [comment]
@am1t We are all talking about the same thing this morning. Browsers don't let you load non https resources when the main page is https. `httpproxy.scripting.com` is http only, and that URL is hardcoded in `readHttpFileThruProxy` function in [this file](http://scripting.com/code/includes/basic/code.js).

The fix is to have that httpproxy server also running on https. Dave just said he has far too many other things to do, I don't blame him. ;-)

**Comment by papascott** [comment]
My two cents worth (as someone who knows just enough JavaScript to be dangerous to himself and others): if we are using our own templates to use with HTTPS, we can load any JavaScript we want. And if there are functions in the standard templates that cause mixed-content errors, we can override those with functions of our own. I'm sure we can figure out how to fix the mixed-content errors without Dave having to change any of his server setups.

**Comment by dcoletta** [comment]
I'm fine with no About tab for the time being, and if I find I need it, I bet I can figure out how to override `readHttpFileThruProxy`. Or, if I can't, then probably that's the meta level at which we would want to make a feature request  - i.e., making sure JS functions that might need to be overriden are overrideable.

**Comment by dcoletta** [comment]
That was part of it. But even after I made the change to the index.html template, the value of `%aboutOutline%` is coming through as `undefined`. 

I can confirm that in the DOM of my page I am picking up the version of code.js you just pushed 20 mins ago. 

What am I missing?

**Comment by scripting** [comment]
@papascott -- i think you;'re saying it's working.

Also at this moment it would be realllllly helpful if people posted links to their blogs.

That is if you want me to help :smile:

**Comment by scripting** [comment]
@sabre23t — that’s correct. It uses a different renderer. You can as you discovered use an html image element.
